src/Data/StringVariants/NonEmptyText.hs

```diff
@@ -82,39 +83,41 @@ nonEmptyTextToText (NonEmptyText t) = t
 -- | Identical to the normal text filter function, but maintains the type-level invariant
 -- that the text length is <= n, unlike unwrapping the text, filtering, then
 -- rewrapping the text.
-filterNonEmptyText :: (Char -> Bool) -> NonEmptyText n -> NonEmptyText n
-filterNonEmptyText f (NonEmptyText t) = NonEmptyText (T.filter f t)
+filterNonEmptyText :: KnownNat n => (Char -> Bool) -> NonEmptyText n -> Maybe (NonEmptyText n)
+filterNonEmptyText f (NonEmptyText t) = mkNonEmptyText (T.filter f t)
 
 -- | Narrows the maximum length, dropping any remaining trailing characters.
-takeNonEmptyText :: forall m n. (KnownNat m, KnownNat n, n <= m) => NonEmptyText m -> NonEmptyText n
+takeNonEmptyText :: forall m n. (KnownNat m, KnownNat n, 1 <= n, n <= m) => NonEmptyText m -> NonEmptyText n
 takeNonEmptyText (NonEmptyText t) =
   if m == n
     then NonEmptyText t
-    else NonEmptyText $ T.take n t
+    -- when the input is stripped, taking from it is guaranteed to be not empty
+    else NonEmptyText $ T.stripEnd $ T.take n t
   where
     m = fromIntegral $ natVal (Proxy @m)
     n = fromIntegral $ natVal (Proxy @n)
 
 -- | Narrows the maximum length, dropping any prefix remaining characters.
-takeNonEmptyTextEnd :: forall m n. (KnownNat m, KnownNat n, n <= m) => NonEmptyText m -> NonEmptyText n
+takeNonEmptyTextEnd :: forall m n. (KnownNat m, KnownNat n, 1 <= n, n <= m) => NonEmptyText m -> NonEmptyText n
 takeNonEmptyTextEnd (NonEmptyText t) =
   if m == n
     then NonEmptyText t
-    else NonEmptyText $ T.takeEnd n t
+    -- when the input is stripped, taking from it is guaranteed to be not empty
+    else NonEmptyText $ T.stripStart $ T.takeEnd n t
   where
     m = fromIntegral $ natVal (Proxy @m)
     n = fromIntegral $ natVal (Proxy @n)
 
 -- | /O(n)/ Splits a 'NonEmptyText' into components of length @chunkSize@. The
--- last element may be shorter than the other chunks, depending on the length
--- of the input.
+-- chunks may be shorter than the chunkSize depending on the length
+-- of the input and spacing. Each chunk is stripped of whitespace.
 chunksOfNonEmptyText ::
   forall chunkSize totalSize.
-  (KnownNat chunkSize, KnownNat totalSize) =>
+  (KnownNat chunkSize, KnownNat totalSize, chunkSize <= totalSize) =>
   NonEmptyText totalSize ->
   [NonEmptyText chunkSize]
 chunksOfNonEmptyText (NonEmptyText t) =
-  map NonEmptyText (T.chunksOf chunkSize t)
+  mapMaybe mkNonEmptyText (T.chunksOf chunkSize t)
```

Comment:
we should validate that this does not break the ACH parsing (the only current usage) before applying it to the backend code.

Reply:
Yeah, the change is that it is going to strip the spaces for the individual chunks. It may break if splitting the ACH into four lines relied on the preserving leading or trailing whitespace. But in that case it's not correct to use the `NonEmptyText` type.

I think that the new semantics is correct. The invariant is that it is impossible to create a NonEmptyText that mkNonEmptyText cannot create. When updating the mwb to the new version, we should check that it still works. It's possible that we may need another function. I'd like to do this change, but then we'll check for breakage when updating string-variants for mwb.